File: Ch1/Java/IsUnique.java
// Implement an algorithm to determine if a string has all unique characters. What if you
// cannot use additional data structures? 
import java.util.HashSet;
public class IsUnique {
    public static void main(String[] args) {
        String[] testStrings = {"IamShrey", "hello", "noduplicates", "shreyamin", "thiscontainsduplicates"};
        for (String testString : testStrings) {
            System.out.printf("Testing: %s\n", testString);
            System.out.printf("\tUsing isUniqueA: %b\n", isUniqueA(testString));
            System.out.printf("\tUsing isUniqueB: %b\n", isUniqueB(testString));
        }
    }

    public static boolean isUniqueA(String chars) {
        HashSet<Character> allChars = new HashSet<Character>();
        for (int i = 0; i < chars.length(); i++) {
            char currChar = chars.charAt(i);
            if (allChars.contains(currChar)) {
                return false;
            }
            allChars.add(currChar);
        }
        return true;
    }

    public static boolean isUniqueB(String chars) {
        if (chars.length() > 128) {
            return false;
        }

        int[] charArray = new int[128];
        for (int i = 0; i < chars.length(); i++) {
            char currChar = chars.charAt(i);
            int currVal = charArray[(int) currChar]++;
            if (currVal > 0) {
                return false;
            }
        }
        return true;
    }
}

File: Ch1/C++/IsUnique.cpp
#include <iostream>
#include <string>
#include <unordered_set>

bool isUniqueA(std::string chars) {
    std::unordered_set<char> allChars;
    for (const auto& currChar : chars) {
        if (allChars.count(currChar)) {
            return false;
        }
        allChars.insert(currChar);
    }
    return true;
}

bool isUniqueB(std::string chars) {
    if (chars.length() > 128) {
        return false;
    }

    int charArray[128] = {0};
    for (const auto& currChar : chars) {
        int currVal = charArray[(int)currChar]++;
        if (currVal > 0) {
            return false;
        }
    }
    return true;
}

int main() {
    std::string testStrings[] = {"IamShrey", "hello", "noduplicates", "shreyamin", "thiscontainsduplicates"};
    for (std::string testString : testStrings) {
        std::cout << "Testing: " << testString << std::endl;
        std::cout << "\tUsing isUniqueA: " << (isUniqueA(testString) ? "true" : "false") << std::endl;
        isUniqueB(testString);
        std::cout << "\tUsing isUniqueB: " << (isUniqueB(testString) ? "true" : "false") << std::endl;
    }
    return 0;
}